eis_toolkit/prediction/fuzzy_overlay.py

```diff
@@ -0,0 +1,112 @@
+        2D Numpy array with the result of the 'or' overlay operation. Values are in range [0, 1].
+
+    Raises:
+        InvalidParameterValueException: If data values are not in range [0, 1].
+    """
+    if any(band_data.min() < 0 or band_data.max() > 1 for band_data in data):
+        raise InvalidParameterValueException("All data must be in range [0, 1]")
+
+    return data.max(axis=0)
+
+
+@beartype
+def product_overlay(data: np.ndarray) -> np.ndarray:
+    """Compute an 'product' overlay operation with fuzzy logic.
+
+    Args:
+        data: The input data as a 3D Numpy array. Each 2D array represents a raster band.
+            Data points should be in the range [0, 1].
+
+    Returns:
+        2D Numpy array with the result of the 'product' overlay operation. Values are in range [0, 1].
+
+    Raises:
+        InvalidParameterValueException: If data values are not in range [0, 1].
+    """
+    if any(band_data.min() < 0 or band_data.max() > 1 for band_data in data):
+        raise InvalidParameterValueException("All data must be in range [0, 1]")
+
+    return np.prod(data, axis=0)
+
+
+@beartype
+def sum_overlay(data: np.ndarray) -> np.ndarray:
+    """Compute an 'sum' overlay operation with fuzzy logic.
+
+    Args:
+        data: The input data as a 3D Numpy array. Each 2D array represents a raster band.
+            Data points should be in the range [0, 1].
+
+    Returns:
+        2D Numpy array with the result of the 'sum' overlay operation. Values are in range [0, 1].
+
+    Raises:
+        InvalidParameterValueException: If data values are not in range [0, 1].
+    """
+    if any(band_data.min() < 0 or band_data.max() > 1 for band_data in data):
+        raise InvalidParameterValueException("All data must be in range [0, 1]")
+
+    return data.sum(axis=0) - np.prod(data, axis=0)
+
+
+@beartype
+def gamma_overlay(data: np.ndarray, gamma: float) -> np.ndarray:
+    """Compute an 'gamma' overlay operation with fuzzy logic.
+
+    Args:
+        data: The input data as a 3D Numpy array. Each 2D array represents a raster band.
+            Data points should be in the range [0, 1].
+        gamma: The gamma parameter. With gamma value 0, result will be same as 'product'overlay.
+            When gamma is closer to 1, the weight of 'sum' overlay is increased.
+            Value must be in the range [0, 1].
+
+    Returns:
+        2D Numpy array with the result of the 'gamma' overlay operation. Values are in range [0, 1].
+
+    Raises:
+        InvalidParameterValueException: If data values or gamma are not in range [0, 1].
+    """
+    if any(band_data.min() < 0 or band_data.max() > 1 for band_data in data):
+        raise InvalidParameterValueException("All data must be in range [0, 1]")
+
+    if gamma < 0 or gamma > 1:
+        raise InvalidParameterValueException("The gamma parameter must be in range [0, 1]")
+
+    fuzzy_sum = data.sum(axis=0) - np.prod(data, axis=0)
+    fuzzy_product = np.prod(data, axis=0)
```

Comment:
Why not reuse already defined functionality for sum and product?

```suggestion
    fuzzy_product = product_overlay(data=data)
```

Reply:
I thought about this briefly, but did not do it because the raster data would be then checked three times while calling gamma_overlay. Do you think the min and max checks are really quick even for large raster data? If so, I will change it as you suggested

Reply:
The checking could be reduced to two times by omitting input data checking from the gamma_overlay function actually

Reply:
True about the performance loss. But I would prioritize in these cases maintainability rather than optimization. You could use memory caching to make sure that same min and max are not calculated multiple times with `functools.cache` (<https://docs.python.org/3/library/functools.html#functools.cache>). I believe it could be used as a decorator for the check function.

Reply:
I tried this, but there seemed to be some problem (said _unhashable type: 'numpy.ndarray'_, I did not investigate further). Anyway, with quite a huge array of 80M cells gamma overlay with two checks was performed in 1s, so I do not think this need optimizing at this point.